examples/with-wallet-stamper/src/components/turnkey-provider.tsx

```diff
@@ -131,37 +212,128 @@ export const TurnkeyProvider: React.FC<TurnkeyProviderProps> = ({
       return [];
     }
-    const { wallets } = await walletClient?.getWallets({
-      organizationId: user?.organizationId,
-    });
+
+    let wallets = [];
+
+    const iframeClient = await getActiveIframeClient();
+
+    if (iframeClient) {
+      const { wallets: walletsResponse } = await iframeClient.getWallets({
+        organizationId: user?.organizationId,
+      });
+      wallets = walletsResponse;
+    } else {
+      const { wallets: walletsResponse } = await walletClient?.getWallets({
+        organizationId: user?.organizationId,
+      });
+      wallets = walletsResponse;
+    }
 
     return wallets as unknown as Wallet[];
   }
 
-  async function signInWithWallet(): Promise<User | null> {
-    const publicKey = (wallet as SolanaWalletInterface)?.recoverPublicKey();
-    const { organizationIds } = await getSubOrgByPublicKey(publicKey);
-    const organizationId = organizationIds[0];
-
+  // Sign in with wallet and create an active session
+  async function signInWithWallet(email?: Email): Promise<User | null> {
     let whoami: User | null = null;
-    if (walletClient) {
-      try {
-        whoami = await walletClient.getWhoami({
-          organizationId,
-        });
-        setUser(whoami);
-        router.push("/dashboard");
-      } catch (e) {
-        console.error(e);
+
+    if (!walletClient) {
+      return null;
+    }
+
+    // If no email is provided, log in with wallet. Will require multiple approvals.
+    if (!email) {
+      const publicKey = (wallet as SolanaWalletInterface)?.recoverPublicKey();
+      const { organizationIds } = await getSubOrgByPublicKey(publicKey); // One wallet request
+      const organizationId = organizationIds[0];
+
+      if (walletClient) {
+        try {
+          // Another wallet request
+          whoami = await walletClient.getWhoami({
+            organizationId,
+          });
+          setUser(whoami);
+          router.push("/dashboard");
+        } catch (e) {
+          console.error(e);
+        }
       }
+
+      return whoami;
     }
+
+    // Create an iframe session. Email is required for the next bit.
+    try {
+      const activityPoller = createActivityPoller({
+        client: walletClient,
+        requestFn: walletClient.createReadWriteSession,
+      });
+
+      const completedActivity = await activityPoller({
+        type: "ACTIVITY_TYPE_CREATE_READ_WRITE_SESSION",
+        timestampMs: new Date().getTime().toString(),
+        organizationId: NEXT_PUBLIC_ORGANIZATION_ID, // This can accept the parent org if we don't know the suborg in advance
+        parameters: {
+          email: email!,
+          targetPublicKey: iframeClient?.iframePublicKey!,
+        },
+      });
+
+      const result = completedActivity.result.createReadWriteSessionResult;
+
+      // Need to save resulting bundle and information.
+      // This is required by the authIframeClient abstraction.
+      localStorage.setItem("CREDENTIAL_BUNDLE", result?.credentialBundle!);
```

Comment:
This is soon to be abstracted away once we have proper support in our higher level SDK packages.